### PR TITLE
Baytrail/Cherrytrail suspend-resume + module reload support

### PR DESCRIPTION
**Pull request by plbossart**
Adding all fixes in one PR to help with testing. This works fine for me on Cyan Chromebook and RT5640 device.

### CODE REVIEW

File: sound/soc/sof/ipc.c [hunk 1]
```diff
@@ -335,21 +335,20 @@ int sof_ipc_tx_message_no_pm(struct snd_sof_ipc *ipc, u32 header,
...
 /* handle reply message from DSP */
-int snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
 {
 	struct snd_sof_ipc_msg *msg = &sdev->ipc->msg;
 
 	if (msg->ipc_complete) {
-		dev_err(sdev->dev, "error: no reply expected, received 0x%x",
+		dev_dbg(sdev->dev,
+			"no reply expected, received 0x%x, will be ignored",
```
Comment (lyakh) [inline sound/soc/sof/ipc.c:344]: @ranj063 Does this ever trigger? I see that the return value of this function is never checked, so this patch's only (potential) run-time effect is if the error message disappears. But isn't it really an error? Shouldn't we at least warn?

**Comment by eballetbo** [comment]
After picking these patches, this PR seems to fix all the issues we saw. 

Tested-by: Enric Balletbo i Serra <enric.balletbo@collabora.com>

Thanks, @plbossart for work on this.

File: sound/soc/sof/ipc.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc.c:344]: I saw this error a couple of times, though less often after applying @keyonjie 's idea of only enabling the DONE interrupt when sending a command.

**Comment by plbossart** [comment]
> After picking these patches, this PR seems to fix all the issues we saw.
> 
> Tested-by: Enric Balletbo i Serra [enric.balletbo@collabora.com](mailto:enric.balletbo@collabora.com)
> 
> Thanks, @plbossart for work on this.

Thanks @eballetbo for testing, added your tag in commit messages

**Comment by cujomalainey** [comment]
I left my Cyan in the office so I can't test easily :( but I am looking forward to trying this code out.

**Comment by plbossart** [comment]
I am going to merge this, I'd like all Baytrail to be handled in the next kernel cycle.

**Comment by youling257** [comment]
test patches, sometimes no sound on Bay trail.
[  332.048218] sof-audio-acpi 80860F28:00: error: ipc timed out for 0x60050000 size 12
[  332.048280] sof-audio-acpi 80860F28:00: error: unexpected fault 0x600a0001 trace 0x00000000
[  332.048287] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x8000000000000000
[  332.048290] sof-audio-acpi 80860F28:00: error: mask host: pending no complete no raw 0x0
[  332.048294] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete yes raw 0x600a0001
[  332.048297] sof-audio-acpi 80860F28:00: error: mask DSP: pending no complete no raw 0x0
[  332.048304]  Low Latency: ASoC: trigger FE cmd: 0 failed: -110
[  332.552324] sof-audio-acpi 80860F28:00: error: ipc timed out for 0x60030000 size 12
[  332.552387] sof-audio-acpi 80860F28:00: error: unexpected fault 0x600a0001 trace 0x00000000
[  332.552393] sof-audio-acpi 80860F28:00: error: ipc host -> DSP: pending no complete no raw 0x8000000000000000
[  332.552397] sof-audio-acpi 80860F28:00: error: mask host: pending no complete no raw 0x0
[  332.552400] sof-audio-acpi 80860F28:00: error: ipc DSP -> host: pending no complete yes raw 0x600a0001
[  332.552403] sof-audio-acpi 80860F28:00: error: mask DSP: pending no complete no raw 0x0

**Comment by plbossart** [comment]
@eballetbo side question on Cyan: do you see "Headphone Jack" and "Headset Mic Jack" controls on your device ('amixer controls' will give you the list)? I did an upgrade to Ubuntu 20.04 and of course the jack detection doesn't work, not sure if I am missing something (ACPI or .config setup)? Thanks!

I added UCM support but that stupid jack is not functional for some reason.

**Comment by youling257** [comment]
> @eballetbo side question on Cyan: do you see "Headphone Jack" and "Headset Mic Jack" controls on your device ('amixer controls' will give you the list)? I did an upgrade to Ubuntu 20.04 and of course the jack detection doesn't work, not sure if I am missing something (ACPI or .config setup)? Thanks!
> 
> I added UCM support but that stupid jack is not functional for some reason.
sof-bytcht-rt5640: reuse legacy settings, Just test if controls for the legacy driver are available

what's the mean of "legacy driver are available"?

(  30.692|   0.006) I: [pulseaudio] (alsa-lib)main.c: error: failed to import hw:0 use case configuration -2
(  30.693|   0.000) I: [pulseaudio] (alsa-lib)main.c: error: failed to import sof-bytcr-rt5640 use case configuration -2
(  30.693|   0.000) I: [pulseaudio] alsa-ucm.c: UCM not available for card sof-bytcr-rt5640

**Comment by youling257** [comment]
@plbossart how to make ucm2 work with sofbytcrrt5640?

**Comment by plbossart** [comment]
> @plbossart how to make ucm2 work with sofbytcrrt5640?

You need the latest alsa-lib, alsa-ucm-conf and the latest topic/sof-dev kernel branch (all code submitted upstream)

**Pull request merged**